### DISCUSSION TITLE
Attempting to transmit out of 2 SX1262s

### DISCUSSION THREAD

**Discussion by lehornbill** [jgromes/RadioLib#1282] [Q&A] [answered]
Hello everyone,

I'm working on a project where I need to connect two or more SX1262 radios to a single SPI peripheral on a Raspberry Pi 4. My goal is to configure and transmit from multiple radios within the same code using the RadioLib library.

I tried to achieve this by initializing 2 radios, and then using the NSS pin to determine which radio should transmit. Using this same script, I could get the individual radios to work by commenting off the other radios, and have not touched the wiring since. Also, I found that only the last initialized radio will transmit successfully, while the other radio would have its pins be busy. Would be grateful if anyone could give some advice; I am new to this scene. Thanks!

Problem:

Single Radio Success: 
Individually, each SX1262 radio works perfectly when using the NonArduino/Raspberry Pi examples from RadioLib.
    
Multiple Radios Issue: 
When I try to initialize and use two radios simultaneously, Only the last initialized radio transmits successfully. The other radio fails to transmit and RadioLib says "GPIO PIN IS BUSY" as shown in image below. After initialization, I can't seem to read from or write to the NSS (Chip Select) pins manually. Whichever radio is initialized last appears to take over the SPI bus.

Testing Individual Radios:
 I commented out one radio in the code to test them individually,  each radio works when tested alone, without changing the wiring.
    
 
How can I properly control and transmit from two or more SX1262 radios connected to a single SPI peripheral on a Raspberry Pi 4 using RadioLib? Is there a way to manage the NSS pins or SPI transactions to prevent conflicts between the radios? I'm relatively new to SPI communication and RadioLib, so any detailed explanations or guidance would be greatly appreciated.

Thank you.

```
#include <RadioLib/RadioLib.h>
#include "PiHal.h"

int main(int argc, char** argv) {
  // Initialize radios
  const uint8_t nss1 = 12;  // NSS for Radio 1
  const uint8_t nss2 = 22;  // NSS for Radio 2
  PiHal* hal = new PiHal(0);
  
  // Create radio instances
  SX1262 radio1 = new Module(hal, nss1, 27, 25, RADIOLIB_NC);
  SX1262 radio2 = new Module(hal, nss2, 23, 24, RADIOLIB_NC);  // Same pins for both radios except NSS

  printf("[SX1262] Initializing Radio 1 ... ");
  int state = radio1.begin();
  if (state != RADIOLIB_ERR_NONE) {
    printf("failed, code %d\n", state);
    return 1;
  }
  printf("success!\n");

  printf("[SX1262] Initializing Radio 2 ... ");
  state = radio2.begin();
  if (state != RADIOLIB_ERR_NONE) {
    printf("failed, code %d\n", state);
    return 1;
  }
  printf("success!\n");
  

  // Initialize NSS pins
  hal->pinMode(nss1, PI_OUTPUT);
  hal->pinMode(nss2, PI_OUTPUT);
  hal->digitalWrite(nss1, LG_HIGH);  // Deselect both radios initially
  hal->digitalWrite(nss2, LG_HIGH);

  int count = 0;
  for (;;) {
    // Send packet from Radio 1
    hal->digitalWrite(nss1, LG_LOW);  // Select Radio 1
    printf("[SX1262] Transmitting packet from Radio 1 ... ");
    char str[64];
    sprintf(str, "Hello from Radio 1! #%d", count++);
    state = radio1.transmit(str);
    if (state == RADIOLIB_ERR_NONE) {
      printf("success!\n");
    } else {
      printf("failed, code %d\n", state);
    }
    hal->digitalWrite(nss1, LG_HIGH);  // Deselect Radio 1

    hal->delay(1000);

    // Send packet from Radio 2
    hal->digitalWrite(nss2, LG_LOW);  // Select Radio 2
    printf("[SX1262] Transmitting packet from Radio 2 ... ");
    sprintf(str, "Hello from Radio 2! #%d", count++);
    state = radio2.transmit(str);
    if (state == RADIOLIB_ERR_NONE) {
      printf("success!\n");
    } else {
      printf("failed, code %d\n", state);
    }
    hal->digitalWrite(nss2, LG_HIGH);  // Deselect Radio 2

    hal->delay(1000);
  }

  return 0;
}

```

The logs:
```
[SX1262] Initializing Radio 1 ... success!
[SX1262] Initializing Radio 2 ... success!
Could not claim pin 12 for mode 1: GPIO busy
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
[SX1262] Transmitting packet from Radio 1 ... failed, code -2
Error writing value to pin 12: GPIO busy
[SX1262] Transmitting packet from Radio 2 ... success!
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
Error writing value to pin 12: GPIO busy
[SX1262] Transmitting packet from Radio 1 ... failed, code -2
Error writing value to pin 12: GPIO busy
[SX1262] Transmitting packet from Radio 2 ... success!
```

**Comment by HeadBoffin** [comment]
There is a topic two down from yours about wiring up two radios, probably best you read that and try some suggestions there first.

**Comment by jgromes** [comment, marked as answer]
Couple of points:

1. You really shouldn't change the NSS pin state. Let the library handle that. Some SPI devices will not work if you hold their NSS pin low for too long, I'm not sure if that is the case for SX126x.
2. SX126x without BUSY pin doesn't work well. RadioLib tries to make it work by adding some delays instead of waiting for BUSY, but that's not really how its supposed to be done.
3. The primary issue seems to be that `Error writing value to pin 12: GPIO busy` - that is not from RadioLib, most likely from the HAL.

**Reply by lehornbill** [reply]
Sflr, I followed your advice and initialized another hal, it works now. Thanks!
